wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,431 @@
+        -- over the wire if any of the inputs it consumes are mentioned in
+        -- the 'Pending' set.
+        dependsOnPendingTx :: ScheduledTx -> Bool
+        dependsOnPendingTx (ScheduledTx _ txAux _) =
+            let txs = pending ^. pendingTransactions . fromDb
+                inputs = List.foldl' updateFn mempty $ (Core.taTx txAux) ^. Core.txInputs . to NonEmpty.toList
+            in any (isJust . flip M.lookup txs) inputs
+
+        updateFn :: [Core.TxId] -> Core.TxIn -> [Core.TxId]
+        updateFn acc (Core.TxInUnknown _ _)   = acc
+        updateFn acc (Core.TxInUtxo txHash _) = txHash : acc
+
+
+-- | Overrides the 'Schedule' with an input transaction and a
+-- custom 'Schedule'. Useful to force dispatching, especially in tests.
+overrideSchedule :: WalletSubmission m
+                 -> Slot
+                 -> ScheduledTx
+                 -> WalletSubmission m
+overrideSchedule ws slot scheduled = ws & over (wsState . wssSchedule) override
+    where
+        override :: Schedule -> Schedule
+        override (Schedule s n) =
+            Schedule (IntMap.insertWith mappend (castSlot slot) [scheduled] s) n
+
+
+-- | A getter to the local pending set stored in this 'WalletSubmission'.
+localPendingSet :: Getter (WalletSubmission m) Pending
+localPendingSet = wsState . wssPendingSet
+
+getCurrentSlot :: Getter (WalletSubmission m) Slot
+getCurrentSlot = wsState . wssCurrentSlot
+
+getSchedule :: Getter (WalletSubmission m) Schedule
+getSchedule = wsState . wssSchedule
+
+
+--
+-- Internal API
+--
+
+-- | Schedule the full list of pending transactions.
+-- The transactions will be scheduled immediately in the next 'Slot'.
+schedulePending :: Pending
+                -> WalletSubmission m
+                -> WalletSubmission m
+schedulePending pending ws =
+    let currentSlot = ws ^. wsState . wssCurrentSlot
+    in ws & over (wsState . wssSchedule) (scheduleNext currentSlot)
+    where
+        scheduleNext :: Slot -> Schedule -> Schedule
+        scheduleNext currentSlot (Schedule s n) =
+            let txs     = map toEntry (pending ^. pendingTransactions . fromDb . to M.toList)
+                toEntry (txId, txAux) = ScheduledTx txId txAux (SubmissionCount 0)
+            in Schedule (IntMap.insertWith mappend (castSlot $ mapSlot succ currentSlot) txs s) n
+
+-- | Purges all the scheduled transactions in the given 'Slot'. This is
+-- automatically called by 'tick', and it's responsibility of @rho@ to
+-- correctly reschedule all the transaction.
+purgeSlot :: Slot -> Schedule -> Schedule
+purgeSlot slot (Schedule s n) = Schedule (IntMap.delete (castSlot slot) s) n
+
+-- Ready-to-use 'ResubmissionFunction's.
+
+-- | A 'RetryPolicy' is simply a function which instruct the 'Schedule' when
+-- to attempt resubmitting the given 'ScheduledTx' item.
+type RetryPolicy = SubmissionCount -> Slot -> Maybe Slot
+
+limited :: MaxRetries -> RetryPolicy -> RetryPolicy
```

Comment:
Maybe add a comment stating that this is an internal function only, and that the policies we provide take `MaxRetries` as an argument to enforce that we _always_ have a maximum?

Reply:
yessir, will do!